# This House Price Prediction model uses a Multi-Layer Perceptron (MLP). 
**I performed EDA and data preprocessing to extract key features. The model is trained with the Adam optimizer and Mean Squared Error (MSE) loss function.**

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/house-price/house_prices.csv


In [2]:
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.metrics import mean_squared_error,r2_score

2025-07-26 14:07:39.114984: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1753538859.361745      13 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1753538859.429398      13 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


In [3]:
data=pd.read_csv("/kaggle/input/house-price/house_prices.csv")

In [4]:
data.tail(20)


/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1458: RuntimeWarning: invalid value encountered in greater
  has_large_values = (abs_vals > 1e6).any()
/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in less
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()
/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in greater
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()
/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1458: RuntimeWarning: invalid value encountered in greater
  has_large_values = (abs_vals > 1e6).any()
/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in less
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()
/usr/local/lib/python3.11/dist-packages/pan

Index                                              Title  \
187511  187511  4 BHK Ready to Occupy Flat for sale in Savitry...   
187512  187512  2 BHK Ready to Occupy Flat for sale in Jaivee ...   
187513  187513  3 BHK Ready to Occupy Flat for sale in Motia C...   
187514  187514        3 BHK Ready to Occupy Flat for sale Gazipur   
187515  187515  3 BHK Ready to Occupy Flat for sale in Savitry...   
187516  187516        2 BHK Ready to Occupy Flat for sale Gazipur   
187517  187517  3 BHK Ready to Occupy Flat for sale in Sushma ...   
187518  187518  3 BHK Ready to Occupy Flat for sale in Sushma ...   
187519  187519        3 BHK Ready to Occupy Flat for sale Dhakoli   
187520  187520  3 BHK Ready to Occupy Flat for sale in Sushma ...   
187521  187521     4 BHK Ready to Occupy Flat for sale Nagla Road   
187522  187522  3 BHK Ready to Occupy Flat for sale in GHB Spl...   
187523  187523  5 BHK Ready to Occupy Flat for sale in Orvis G...   
187524  187524  3 BHK Ready to Occupy Flat for sale in Sushma ...   
187525  187525  3 BHK Ready to Occupy Flat for sale in Peer Mu...   
187526  187526  3 BHK Ready to Occupy Flat for sale in Bollywo...   
187527  187527  3 BHK Ready to Occupy Flat for sale in Sushma ...   
187528  187528  3 BHK Ready to Occupy Flat for sale in Bollywo...   
187529  187529  2 BHK Ready to Occupy Flat for sale in Friends...   
187530  187530  3 BHK Ready to Occupy Flat for sale in Affinit...   

                                              Description Amount(in rupees)  \
187511  This exquisite 4 BHK Flat is offered for sale ...           60 Lac    
187512  Peer Muchalla, Zirakpur has an appealing 2 BHK...           40 Lac    
187513  This ready to move-in 3 BHK flat is available ...    Call for Price   
187514  This lovely 3 BHK Flat in Gazipur, Zirakpur is...           65 Lac    
187515  This exquisite 3 BHK Flat is offered for sale ...           55 Lac    
187516  This beautiful 2 BHK Flat is available for sal...         29.9 Lac    
187517  One can find this stunning 3 BHK flat for sale...         83.5 Lac    
187518  This ready to move-in 3 BHK flat is available ...         54.9 Lac    
187519  This magnificent 3 BHK Flat is available for s...           37 Lac    
187520  This lovely 3 BHK Flat in Gazipur, Zirakpur is...           78 Lac    
187521  Nagla Road, Zirakpur has an appealing 4 BHK fl...          1.18 Cr    
187522  Up for immediate sale is a 3 BHK apartment in ...           80 Lac    
187523  This magnificent 5 BHK Flat is available for s...          1.19 Cr    
187524  One can find this stunning 3 BHK flat for sale...         69.6 Lac    
187525  Have a look at this immaculate 3 BHK flat for ...         44.9 Lac    
187526  This magnificent 3 BHK Flat is available for s...           63 Lac    
187527  Have a look at this immaculate 3 BHK flat for ...           55 Lac    
187528  Gazipur, Zirakpur has an appealing 3 BHK flat ...           76 Lac    
187529  Up for immediate sale is a 2 BHK apartment in ...           30 Lac    
187530  This exquisite 3 BHK Flat is offered for sale ...          1.18 Cr    

        Price (in rupees)  location Carpet Area         Status        Floor  \
187511             4594.0  zirakpur         NaN  Ready to Move   2 out of 2   
187512             3810.0  zirakpur    950 sqft  Ready to Move   3 out of 3   
187513                NaN  zirakpur   1200 sqft  Ready to Move   3 out of 3   
187514             3939.0  zirakpur   1200 sqft  Ready to Move   1 out of 4   
187515             3868.0  zirakpur   1200 sqft  Ready to Move   2 out of 3   
187516             3322.0  zirakpur    750 sqft  Ready to Move   2 out of 3   
187517             4898.0  zirakpur   1195 sqft  Ready to Move   5 out of 9   
187518             3268.0  zirakpur         NaN  Ready to Move   7 out of 8   
187519             4405.0  zirakpur         NaN  Ready to Move   3 out of 4   
187520             4615.0  zirakpur   1352 sqft  Ready to Move  2 out of 12   
187521             5816.0  zir

In [5]:
data.describe()

/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1458: RuntimeWarning: invalid value encountered in greater
  has_large_values = (abs_vals > 1e6).any()
/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in less
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()
/usr/local/lib/python3.11/dist-packages/pandas/io/formats/format.py:1459: RuntimeWarning: invalid value encountered in greater
  has_small_values = ((abs_vals < 10 ** (-self.digits)) & (abs_vals > 0)).any()


Index  Price (in rupees)  Dimensions  Plot Area
count  187531.000000       1.698660e+05         0.0        0.0
mean    93765.000000       7.583772e+03         NaN        NaN
std     54135.681003       2.724171e+04         NaN        NaN
min         0.000000       0.000000e+00         NaN        NaN
25%     46882.500000       4.297000e+03         NaN        NaN
50%     93765.000000       6.034000e+03         NaN        NaN
75%    140647.500000       9.450000e+03         NaN        NaN
max    187530.000000       6.700000e+06         NaN        NaN

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 187531 entries, 0 to 187530
Data columns (total 21 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Index              187531 non-null  int64  
 1   Title              187531 non-null  object 
 2   Description        184508 non-null  object 
 3   Amount(in rupees)  187531 non-null  object 
 4   Price (in rupees)  169866 non-null  float64
 5   location           187531 non-null  object 
 6   Carpet Area        106858 non-null  object 
 7   Status             186916 non-null  object 
 8   Floor              180454 non-null  object 
 9   Transaction        187448 non-null  object 
 10  Furnishing         184634 non-null  object 
 11  facing             117298 non-null  object 
 12  overlooking        106095 non-null  object 
 13  Society            77853 non-null   object 
 14  Bathroom           186703 non-null  object 
 15  Balcony            138596 non-null  object 
 16  Ca

**There are many missing values in this data set we can use the feature which has fever number of missing values.**

In [7]:
print(data.isna().sum())  

Index                     0
Title                     0
Description            3023
Amount(in rupees)         0
Price (in rupees)     17665
location                  0
Carpet Area           80673
Status                  615
Floor                  7077
Transaction              83
Furnishing             2897
facing                70233
overlooking           81436
Society              109678
Bathroom                828
Balcony               48935
Car Parking          103357
Ownership             65517
Super Area           107685
Dimensions           187531
Plot Area            187531
dtype: int64


**To better understand the nature of categorical and discrete features, I analyzed the unique values present in several columns.**

In [8]:
a=["location","Status","Floor","Furnishing","facing","overlooking","Bathroom","Balcony","Car Parking","Ownership"]
for cat in a :
    print(f"{cat}:\n{data[cat].unique()}\n")


location:
['thane' 'navi-mumbai' 'nagpur' 'mumbai' 'ahmedabad' 'bangalore' 'chennai'
 'gurgaon' 'hyderabad' 'indore' 'jaipur' 'kolkata' 'lucknow' 'new-delhi'
 'noida' 'pune' 'agra' 'ahmadnagar' 'allahabad' 'aurangabad' 'badlapur'
 'belgaum' 'bhiwadi' 'bhiwandi' 'bhopal' 'bhubaneswar' 'chandigarh'
 'coimbatore' 'dehradun' 'durgapur' 'ernakulam' 'faridabad' 'ghaziabad'
 'goa' 'greater-noida' 'guntur' 'guwahati' 'gwalior' 'haridwar' 'jabalpur'
 'jamshedpur' 'jodhpur' 'kalyan' 'kanpur' 'kochi' 'kozhikode' 'ludhiana'
 'madurai' 'mangalore' 'mohali' 'mysore' 'nashik' 'navsari' 'nellore'
 'palakkad' 'palghar' 'panchkula' 'patna' 'pondicherry' 'raipur'
 'rajahmundry' 'ranchi' 'satara' 'shimla' 'siliguri' 'solapur' 'sonipat'
 'surat' 'thrissur' 'tirupati' 'trichy' 'trivandrum' 'udaipur' 'udupi'
 'vadodara' 'vapi' 'varanasi' 'vijayawada' 'visakhapatnam' 'vrindavan'
 'zirakpur']

Status:
['Ready to Move' nan]

Floor:
['10 out of 11' '3 out of 22' '10 out of 29' '1 out of 3' '20 out of 42'
 '2 out

In [9]:
import re

**Since the ‘Title’ column includes BHK information, a key factor affecting house price, we can extract and convert it into a numerical feature for better model performance.**

In [10]:
data["Title"]=data["Title"].apply(
    lambda x: int(re.search(r'(\d+)\s*BHK', x).group(1)) 
    if re.search(r'(\d+)\s*BHK', x) 
    else None
)


**We can fill the missing values in the ‘Title’ column with 2, as most of the houses are 2 BHK. Using 2 instead of 0 avoids introducing significant distortion, since 0 BHK is unrealistic and would negatively affect the model.**

In [11]:
data["Title"]=data["Title"].fillna(2)
print(data["Title"].isnull().sum())
print(data["Title"].unique())
print(data['Title'].value_counts())
print(data["Title"].isnull().sum())

0
[ 1.  2.  3.  4.  5.  6.  8.  7. 10.  9.]
Title
3.0     80848
2.0     76429
4.0     16600
1.0     12284
5.0      1183
6.0        98
10.0       43
7.0        22
8.0        17
9.0         7
Name: count, dtype: int64
0


**We convert house price values written in words (e.g., ‘50 Lakhs’, ‘1.2 Crores’) into their corresponding numerical format (e.g., 5000000, 12000000) to enable accurate processing and modeling.**

In [12]:
def convert_amount(value):
    try:
        value = str(value).strip()
        if 'Cr' in value:
            return float(value.replace('Cr', '').strip()) * 1e7
        elif 'Lac' in value:
            return float(value.replace('Lac', '').strip()) * 1e5
        elif value.replace('.', '', 1).isdigit():
            return float(value)
        else:
            return None  # For cases like 'Call for Price'
    except:
        return None


data['Amount(in rupees)'] = data['Amount(in rupees)'].apply(convert_amount)


In [13]:
data['Amount(in rupees)'] = data['Amount(in rupees)'].replace('Call for Price', np.nan)

In [14]:
pd.set_option('display.float_format', '{:,.0f}'.format)

In [15]:
print(data.groupby("Title")["Amount(in rupees)"].mean())



Title
1      3,531,214
2      6,219,479
3     13,784,577
4     30,646,983
5     46,354,768
6     70,386,957
7     56,895,238
8    208,985,000
9     36,357,143
10    86,305,128
Name: Amount(in rupees), dtype: float64


**Here, I am replacing the missing price values with the average amount of houses having the same BHK. This helps in estimating a price closer to the actual value.**

In [16]:
data["Amount(in rupees)"] = data["Amount(in rupees)"].fillna(
    data.groupby("Title")["Amount(in rupees)"].transform("mean")
)

In [17]:
print(data["Amount(in rupees)"].isnull().sum())

0



**Here, I am replacing the missing values in the ‘location’ column by using the average house price for each location. First, I calculate the average price per location using groupby. Then I map this average to the ‘location’ column. Finally, if any values are still missing, I fill them with the overall average price.Since the ‘location’ column contains many unique place names, using one-hot encoding would create a large number of features, which may lead to overfitting and increase the model complexity. Therefore, replacing the location with its corresponding average house price is a better approach in this case, as it helps convert categorical data into meaningful numerical values while keeping the feature space compact and avoiding overfitting.**


In [18]:
city_avg_price = data.groupby('location')['Amount(in rupees)'].mean()
data['location'] = data['location'].map(city_avg_price)
data['location'].fillna(data['location'].mean())

0        9,623,260
1        9,623,260
2        9,623,260
3        9,623,260
4        9,623,260
            ...   
187526   6,988,955
187527   6,988,955
187528   6,988,955
187529   6,988,955
187530   6,988,955
Name: location, Length: 187531, dtype: float64

In [19]:
print(data["location"].isnull().sum())

0


In [20]:
data["Carpet Area"] = data["Carpet Area"].str.replace("sqft", "", regex=False).str.strip()
data["Carpet Area"] = pd.to_numeric(data["Carpet Area"], errors="coerce")


data["Carpet Area"] = data["Carpet Area"].fillna(
    data.groupby("Title")["Carpet Area"].transform("mean")
)

data["Carpet Area"]=data["Carpet Area"].fillna(data["Carpet Area"].mean())
print("Missing Carpet Area values:", data["Carpet Area"].isnull().sum())

Missing Carpet Area values: 0


**In this step, we convert the textual floor information, such as “14 out of 38” or “Ground out of 10”, into a normalized numerical value between 0 and 1 to make it suitable for machine learning models. The transformation begins by splitting each floor entry into two parts: the current floor level and the total number of floors in the building. For cases where the current floor is described using non-numeric labels like “Ground”, “Upper Basement”, or “Lower Basement”, we treat them as floor number 0. Once both the floor number and total floors are extracted and converted to integers, we calculate the normalized floor position by dividing the current floor number by the total number of floors. This results in a value that represents the relative height of the property within the building. If any error occurs during parsing, we assign a NaN value, which is later replaced with 0, assuming it refers to a ground-level or basement unit. This approach ensures that all floor information is numerically consistent and scaled, enhancing the model’s ability to learn from this feature.**

In [21]:
def convert_floor(value):
    try:
        parts = value.strip().split(" out of ")
        floor_part = parts[0].strip().lower()
        total_floors = int(parts[1])

       
        if "ground" in floor_part:
            floor_num = 0
        elif "upper basement" in floor_part:
            floor_num = 0
        elif "lower basement" in floor_part:
            floor_num = 0
        else:
            floor_num = int(floor_part)

        
        normalized_value = floor_num / total_floors
        return max(0, min(normalized_value, 1))  # clamp to [0, 1]
    except:
        return np.nan  # Invalid data fallback


data["Floor"] = data["Floor"].apply(convert_floor)
data["Floor"] = data["Floor"].fillna(0)

print(data["Floor"].head())
print(data["Floor"].unique())

print("Missing Floor values:", data["Floor"].isnull().sum())

0   1
1   0
2   0
3   0
4   0
Name: Floor, dtype: float64
[0.90909091 0.13636364 0.34482759 0.33333333 0.47619048 0.28571429
 0.8        0.         0.11111111 0.3        0.66666667 0.4
 0.625      0.71428571 1.         0.42857143 0.48387097 0.07142857
 0.85714286 0.375      0.64705882 0.59259259 0.48275862 0.44444444
 0.03703704 0.5        0.42105263 0.75       0.11764706 0.15789474
 0.21428571 0.25       0.4375     0.58333333 0.82352941 0.77777778
 0.23333333 0.6        0.08333333 0.125      0.14285714 0.15384615
 0.05882353 0.88235294 0.88888889 0.2962963  0.80769231 0.41666667
 0.16666667 0.35714286 0.03448276 0.3125     0.10714286 0.2
 0.83333333 0.17857143 0.15       0.46666667 0.46153846 0.06896552
 0.57142857 0.41935484 0.13043478 0.61111111 0.10526316 0.69565217
 0.1        0.62962963 0.42307692 0.46875    0.20689655 0.05
 0.875      0.63636364 0.71875    0.54545455 0.96296296 0.47368421
 0.63333333 0.63157895 0.32258065 0.72727273 0.79166667 0.36842105
 0.60869565 0.13333333 0

In [22]:
data['Furnishing'] = data['Furnishing'].fillna('Unknown')
data['facing'] = data['facing'].fillna('Unknown')


data = pd.get_dummies(data, columns=['Furnishing', 'facing'], prefix=['Furnish', 'Face'])

In [23]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 187531 entries, 0 to 187530
Data columns (total 32 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   Index                   187531 non-null  int64  
 1   Title                   187531 non-null  float64
 2   Description             184508 non-null  object 
 3   Amount(in rupees)       187531 non-null  float64
 4   Price (in rupees)       169866 non-null  float64
 5   location                187531 non-null  float64
 6   Carpet Area             187531 non-null  float64
 7   Status                  186916 non-null  object 
 8   Floor                   187531 non-null  float64
 9   Transaction             187448 non-null  object 
 10  overlooking             106095 non-null  object 
 11  Society                 77853 non-null   object 
 12  Bathroom                186703 non-null  object 
 13  Balcony                 138596 non-null  object 
 14  Car Parking         

In [24]:

data['overlooking'] = data['overlooking'].fillna('')

data['overlooking'] = data['overlooking'].apply(lambda x: [i.strip() for i in x.split(',')] if x else [])

from sklearn.preprocessing import MultiLabelBinarizer
mlb = MultiLabelBinarizer()
overlook_encoded = pd.DataFrame(mlb.fit_transform(data['overlooking']), columns=['Overlook_' + c for c in mlb.classes_])

data = pd.concat([data.drop(columns=['overlooking']), overlook_encoded], axis=1)

In [25]:
data['Bathroom'] = data['Bathroom'].replace('> 10', '11')
data['Bathroom'] = pd.to_numeric(data['Bathroom'], errors='coerce')
data['Bathroom']=data['Bathroom'].fillna(data['Bathroom'].median())

In [26]:
data['Balcony'] = data['Balcony'].replace('> 10', '11')
data['Balcony'] = pd.to_numeric(data['Balcony'], errors='coerce')
data['Balcony']=data['Balcony'].fillna(data['Balcony'].median())

In [27]:
def extract_parking_num(x):
    if pd.isna(x):
        return None
    x = str(x) 
    match = re.search(r'\d+', x)
    return int(match.group()) if match else None

data['Car Parking'] = data['Car Parking'].apply(extract_parking_num)
data['Car Parking'] = data['Car Parking'].fillna(data['Car Parking'].median())

In [28]:
data['Ownership'].fillna('Unknown')
data = pd.get_dummies(data, columns=['Ownership'], prefix='Owner')

In [29]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 187531 entries, 0 to 187530
Data columns (total 38 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   Index                       187531 non-null  int64  
 1   Title                       187531 non-null  float64
 2   Description                 184508 non-null  object 
 3   Amount(in rupees)           187531 non-null  float64
 4   Price (in rupees)           169866 non-null  float64
 5   location                    187531 non-null  float64
 6   Carpet Area                 187531 non-null  float64
 7   Status                      186916 non-null  object 
 8   Floor                       187531 non-null  float64
 9   Transaction                 187448 non-null  object 
 10  Society                     77853 non-null   object 
 11  Bathroom                    187531 non-null  float64
 12  Balcony                     187531 non-null  float64
 13  Car Parking   

In [30]:

feature_columns = [
    'Title',
    'Carpet Area',
    'Floor',
    'Bathroom',
    'Balcony',
    'Car Parking',
    'Furnish_Furnished',
    'Furnish_Semi-Furnished',
    'Furnish_Unfurnished',
    'Furnish_Unknown',
    'Face_East',
    'Face_North',
    'Face_North - East',
    'Face_North - West',
    'Face_South',
    'Face_South - East',
    'Face_South -West',
    'Face_Unknown',
    'Face_West',
    'Overlook_Garden/Park',
    'Overlook_Main Road',
    'Overlook_Not Available',
    'Overlook_Pool',
    'Owner_Co-operative Society',
    'Owner_Freehold',
    'Owner_Leasehold',
    'Owner_Power Of Attorney'
]


x_data = data[feature_columns]
y_data = data['Amount(in rupees)']

In [31]:
print(x_data.info())
print(y_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 187531 entries, 0 to 187530
Data columns (total 27 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   Title                       187531 non-null  float64
 1   Carpet Area                 187531 non-null  float64
 2   Floor                       187531 non-null  float64
 3   Bathroom                    187531 non-null  float64
 4   Balcony                     187531 non-null  float64
 5   Car Parking                 187531 non-null  float64
 6   Furnish_Furnished           187531 non-null  bool   
 7   Furnish_Semi-Furnished      187531 non-null  bool   
 8   Furnish_Unfurnished         187531 non-null  bool   
 9   Furnish_Unknown             187531 non-null  bool   
 10  Face_East                   187531 non-null  bool   
 11  Face_North                  187531 non-null  bool   
 12  Face_North - East           187531 non-null  bool   
 13  Face_North - W

In [32]:
X_train,X_test,y_train,y_test=train_test_split(x_data,y_data,test_size=0.2,random_state=34)
X_train= X_train.astype(int)
X_test=X_test.astype(int)
y_train=y_train.astype(int)
y_test=y_test.astype(int)

In [33]:
print(X_train.head())
print(X_test.head())
print(y_train.head())

        Title  Carpet Area  Floor  Bathroom  Balcony  Car Parking  \
136353      3         1250      0         2        1            1   
137839      3         1760      0         4        3            1   
128582      1          416      0         1        1            1   
2887        2          500      0         2        1            1   
7912        3         1388      0         3        3            1   

        Furnish_Furnished  Furnish_Semi-Furnished  Furnish_Unfurnished  \
136353                  0                       1                    0   
137839                  0                       1                    0   
128582                  0                       0                    1   
2887                    0                       0                    1   
7912                    0                       0                    1   

        Furnish_Unknown  ...  Face_Unknown  Face_West  Overlook_Garden/Park  \
136353                0  ...             0          0        

In [34]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)
y_train_log = np.log1p(y_train)
y_test_log = np.log1p(y_test)

In [35]:
X_train_scaled[:1]

array([[ 0.55276599,  0.01792145, -0.3764773 , -0.48941226, -1.2168155 ,
        -0.06354929, -0.34712179,  1.05937735, -0.82626476, -0.1256841 ,
        -0.64281744, -0.31113175,  2.60363342, -0.14606743, -0.159874  ,
        -0.11894742, -0.10519521, -0.77348458, -0.21898327, -0.78787193,
        -0.85173065, -0.00577314, -0.37345927, -0.13682507,  0.81846409,
        -0.17078887, -0.07553003]])

In [36]:
print(y_train_log)

136353   16
137839   17
128582   14
2887     15
7912     16
         ..
38435    16
82244    18
153449   16
43498    16
103546   16
Name: Amount(in rupees), Length: 150024, dtype: float64


In [37]:
print("X_Train Shape",X_train_scaled.shape)
print("X_test Shape",X_test_scaled.shape)
print("y_train shape",y_train_log.shape)
print("y_test shape",y_test_log.shape)

X_Train Shape (150024, 27)
X_test Shape (37507, 27)
y_train shape (150024,)
y_test shape (37507,)


In [38]:
model = Sequential([
    Dense(128, activation='relu', input_shape=(27,)),
    Dense(64, activation='relu'),
    Dense(32, activation='relu'),
    Dense(1)  
])
model.compile(optimizer='adam', loss='mse')

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
2025-07-26 14:08:00.705005: E external/local_xla/xla/stream_executor/cuda/cuda_driver.cc:152] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)


In [39]:
model.fit(X_train_scaled, y_train_log, epochs=50, batch_size=27, validation_split=0.1,verbose=1)

Epoch 1/50
5001/5001 ━━━━━━━━━━━━━━━━━━━━ 11s 2ms/step - loss: 9.8748 - val_loss: 0.2380
Epoch 2/50
5001/5001 ━━━━━━━━━━━━━━━━━━━━ 9s 2ms/step - loss: 0.4211 - val_loss: 0.2071
Epoch 3/50
5001/5001 ━━━━━━━━━━━━━━━━━━━━ 9s 2ms/step - loss: 0.2314 - val_loss: 0.1959
Epoch 4/50
5001/5001 ━━━━━━━━━━━━━━━━━━━━ 9s 2ms/step - loss: 0.2334 - val_loss: 0.2551
Epoch 5/50
5001/5001 ━━━━━━━━━━━━━━━━━━━━ 9s 2ms/step - loss: 0.2064 - val_loss: 0.2122
Epoch 6/50
5001/5001 ━━━━━━━━━━━━━━━━━━━━ 9s 2ms/step - loss: 0.2090 - val_loss: 0.1887
Epoch 7/50
5001/5001 ━━━━━━━━━━━━━━━━━━━━ 9s 2ms/step - loss: 0.2910 - val_loss: 0.1861
Epoch 8/50
5001/5001 ━━━━━━━━━━━━━━━━━━━━ 9s 2ms/step - loss: 0.1986 - val_loss: 0.2178
Epoch 9/50
5001/5001 ━━━━━━━━━━━━━━━━━━━━ 9s 2ms/step - loss: 0.1955 - val_loss: 0.1744
Epoch 10/50
5001/5001 ━━━━━━━━━━━━━━━━━━━━ 9s 2ms/step - loss: 0.1849 - val_loss: 0.1873
Epoch 11/50
5001/5001 ━━━━━━━━━━━━━━━━━━━━ 9s 2ms/step - loss: 0.1815 - val_loss: 0.1694
Epoch 12/50
5001/5001 ━━━━━━━

In [40]:
y_pred = model.predict(X_test_scaled)

1173/1173 ━━━━━━━━━━━━━━━━━━━━ 1s 979us/step


In [41]:
mse = mean_squared_error(y_test_log, y_pred)
r2 = r2_score(y_test_log, y_pred)

print("Mean Squared Error:", mse)
print("R² Score:", r2)

Mean Squared Error: 0.15941792027636395
R² Score: 0.7587572614569711



**We have attained a Mean Squared Error (MSE) of 0.1551 and an R² Score of 0.7651, which indicates that our linear regression model is performing reasonably well given the current level of data preprocessing and feature engineering. An R² score of approximately 0.76 suggests that the model is able to explain about 76% of the variance in the target variable, which is a good starting point for a basic model built with minimal processing.**

**However, this performance has been achieved using only simple data cleaning techniques and without extensive optimization or transformation of features. For example, we performed basic cleaning such as converting textual price formats into numeric values, extracting numerical insights from floor information, and filling missing values with location-based or global averages. We have not yet incorporated more advanced steps such as outlier removal, polynomial feature creation, feature selection, encoding location semantics more efficiently, or hyperparameter tuning.**

**This experiment primarily serves as a baseline to understand how a simple Multi-Layer Perceptron (MLP) architecture combined with a linear regression layer behaves on structured tabular data. The encouraging performance highlights the potential of the model and dataset — suggesting that with more refined feature engineering, such as handling skewed distributions, encoding categorical features in a more informative way (e.g., target encoding or embeddings), or switching to a more complex model like XGBoost or a deep MLP with non-linear activations, we can likely achieve much better results.**

**In summary, this project successfully demonstrates the foundational pipeline of cleaning real estate data and applying a basic MLP regression model. The results show promise, and with further refinement in data preprocessing and model optimization, there is significant scope to improve the prediction accuracy and build a robust, production-grade model.**
